src/event-stream/event-stream.service.ts

```diff
@@ -0,0 +1,192 @@
+import { HttpService, Injectable, Logger } from '@nestjs/common';
+import * as WebSocket from 'ws';
+import {
+  Event,
+  EventStream,
+  EventStreamReply,
+  EventStreamSubscription,
+} from './event-stream.interfaces';
+
+const RECONNECT_TIME = 5000;
+const PING_TIMEOUT = 60000;
+
+export class EventStreamSocket {
+  private readonly logger = new Logger(EventStreamSocket.name);
+
+  private ws: WebSocket;
+  private pingTimeout: NodeJS.Timeout;
+  private disconnectDetected = false;
+  private closeRequested = false;
+
+  constructor(
+    private url: string,
+    private topic: string,
+    private handleEvents: (events: Event[]) => void,
+    private handleReceipt: (receipt: EventStreamReply) => void,
+  ) {
+    this.init();
+  }
+
+  private init() {
+    this.disconnectDetected = false;
+    this.closeRequested = false;
+
+    this.ws = new WebSocket(this.url);
+    this.ws
+      .on('open', () => {
+        if (this.disconnectDetected) {
+          this.disconnectDetected = false;
+          this.logger.log('Event stream websocket restored');
+        } else {
+          this.logger.log('Event stream websocket connected');
+        }
+        this.produce({ type: 'listen', topic: this.topic });
+        this.produce({ type: 'listenreplies' });
+        this.ping();
+      })
+      .on('close', () => {
+        if (this.closeRequested) {
+          this.logger.log('Event stream websocket closed');
+        } else {
+          this.disconnectDetected = true;
+          this.logger.error(
+            `Event stream websocket disconnected, attempting to reconnect in ${RECONNECT_TIME}ms`,
+          );
+          setTimeout(() => this.init(), RECONNECT_TIME);
+        }
+      })
+      .on('message', (message: string) => {
+        this.handleMessage(JSON.parse(message));
+      })
+      .on('pong', () => {
+        this.ping();
```

Comment:
Won't this cause us to ping immediately every time we get a pong? (don't see a delay before we emit the `ping` in the method)

Reply:
Certainly looks like it. I based it on the [code from KAT](https://github.com/hyperledger-labs/firefly/blob/main/kat/src/clients/event-streams.ts), so if we decide this requires a change, perhaps that does too.